man/walbc.1.ronn
walbc(1) -- control WalB server processes
===========================================================

## SYNOPSIS

`walbc` [<opt1>] <command> [<args>] [<opt2>]


## DESCRIPTION

**walbc** controlls WalB server processes.

walbc uses TCP connections and talks with server processes
by its original protocols based on cybozulib serializer.


## COMMON OPTIONS (opt1)

* `-a` <ADDR>:
   host name or address (must)
* `-p` <PORT>:
   port number (must)
* `-debug`:
   put debug message.
* `-to` <TIMEOUT>:
   socket timeout [sec].
* `-id` <ID>:
   controller identfier
* `-h`:
   show help message


## COMMAND LIST

* `status` [<VOLUME>]:
  print human-readable status.

* `init-vol` <VOLUME> [<WDEV_PATH>]:
  initialize a volume.

* `clear-vol` <VOLUME>:
  clear a volume.

* `reset-vol` <VOLUME> [<GID>]:
  reset a volume in a server.

* `full-bkp` <VOLUME> [<BULK_LB>]:
  execute full-backup.

* `hash-bkp` <VOLUME> [<BULK_LB>]:
  execute hash-backup.

* `restore` <VOLUME> <GID>:
  restore a volume in an archive server.

* `del-restored` <VOLUME> <GID>:
  delete a restored volume.

* `start` <VOLUME> [<ROLE>]:
  start a volume in a server.

* `stop` <VOLUME> [<MODE>]:
  stop a volume in a server.

* `archive-info` <CMD> <VOLUME> [<PARAM>...]:
  control archive information in a proxy.

* `snapshot` <VOLUME>:
  take a snapshot in a storage.

* `disable-snapshot` <VOLUME> <GID> [<GID>...]:
  disable a snapshot in an archive.

* `enable-snapshot` <VOLUME> <GID> [<GID>...]:
   enable a snapshot in an archive.

* `replicate` <VOLUME> <TYPE> <PARAM> <HOST_INFO_REPL>:
  replicate a volume from an archive to another archive.

* `apply` <VOLUME> <GID>:
  apply old wdiff files to the base image of a volume in an archive.

* `merge` <VOLUME> <GID0> <GID1>:
  merge wdiff files for a volume in an archive.

* `resize` <VOLUME> <SIZE_LB> [zeroclear]:
  resize a volume in a storage or an archive.

* `kick` [<VOLUME>] [<ARCHIVE_ID>]:
  kick background tasks if necessary.

* `bhash` <VOLUME> <GID> [<BULK_LB>]:
  calculate block hash of a volume in an archive.

* `exec` [<ARGUMENT>...]:
  execute a command-line at a server's side.

* `shutdown` [<MODE>]:
  shutdown a server process.

* `get` <TYPE> [<PARAM>...]:
  get some information from a server.



## COMMON ARGUMENTS

* <VOLUME>:
  volume identifier.

* <WDEV_PATH>:
  WalB block device path.

* <GID>:
  Generation ID. This is clean snapshot identifier.


## COMMAND init-vol

Parameter <WDEV_PATH> is required only for `walb-storage` server processes.

## COMMAND reset-vol

Default value of parameter <GID> is 0.

## COMMAND full-bkp/hash-bkp

This is effective for `walb-storage` only.
Parameter <BULK_LB> is bulk size [logical block]. Default value is `DEFAULT_BULK_LB`.

## COMMAND start

<ROLE> is required for `walb-storage` server process.
Specify `target` or `standby` for <ROLE>.

## COMMAND stop

<MODE> is one of `graceful`, `force`, and `empty`.
`empty` is available only for `walb-proxy` server processes.

`empty` mode will wait for all wdiff files being transferred and empty for the volume.

## COMMAND archive-info

This is effective for `walb-proxy` only.
<CMD> is one of `add`, `update`, `get`, `delete`, and `list`.

* `list` <VOLUME>:
  get archive id list.

* `add` <VOLUME> <ARCHIVE_ID> <HOST_INFO_BKP>:
  add an archive info.

* `update` <VOLUME> <ARCHIVE_ID> <HOST_INFO_BKP>:
  update an archive info.

* `get` <VOLUME> <ARCHIVE_ID>:
  get an archive info.

* `delete` <VOLUME> <ARCHIVE_ID>:
  delete an archive info.


<ARCHIVE_ID> is identifier of an archive server.
<HOST_INFO_BKP> is `ADDR:PORT COMPRESS_OPT DELAY` string.
<ADDR> is hostname or IP address.
<PORT> is listen port.
<COMPRESS_OPT> is `TYPE:LEVEL:NR_CPU` string.
<TYPE> is `snappy`, `gzip`, `lzma`, or `none`.
<LEVEL> is compression level from 0 to 9.
<NR_CPU> is number of CPU cores to use for wdiff compression.
<DELAY> is wdiff transfer delay in seconds: from wlog received to wdiff transferring.

<COMPRESS_OPT> and <DELAY> can be omitted.


## COMMAND snapshot

This is available for `walb-storage` only.

## COMMAND disable-snapshot/enable-snapshot

This is available for `walb-archive` only.
You can specify multiple GIDs.

## COMMAND replicate

<TYPE> is `size` or `gid` string.
If <TYPE> is `size`, <PARAM> is wdiff amount [MiB] that is allowed to remain.
Otherwise, <PARAM> is gid that is replication target.

<HOST_INFO_REPL> is `ADDR:PORT DO_RESYNC COMPRESS_OPT MAX_WDIFF_MERGE_SIZE BULK_LB` string.
See `archive-info` command section for <ADDR:PORT>, <COMPRESS_OPT>,
and `full-bkp` command section for <BULK_LB> arguments.
<DO_RESYNC> is 0 or 1. 1 will enable to execute resync-replication.
<MAX_WDIFF_MERGE_SIZE> is maximum size of wdiff files to merge at once [bytes].
Size suffix `K`, `M`, `G` are available.

<DO_RESYNC> and succeeding arguments can be omitted.

## COMMAND resize

This is available for `walb-storage` and `walb-archive`.

<SIZE_LB> is new size [logical block].
You can specify 0 for `walb-storage` to auto-detect the underlying data device.
You must resize the underlying data device by yourself before.

Specify `zeroclear` to zero-clear the extended area.
This is available for `walb-archive` only.

## COMMAND kick

For `walb-storage`, any parameter is not required.
This kicks wlog-transfer tasks.

For `walb-proxy`, you can specify <VOLUME> or <VOLUME>, <ARCHIVE_ID> pair.
No parameter is also accepted.
This kicks wdiff-transfer tasks.

## COMMAND exec

Specify full path of the executable, files and directories
because the server process's current directory is unkown.

## COMMAND shutdown

<MODE> is `graceful` or `force`.
Default value is `gracefull`.

There is another method to shutdown server processes forcely
by sending `SIGTERM` signal.
It works as the command `shutdown force`.


## COMMAND get

For any server processes,

* `get state` <VOLUME>:
  get state string of a volume.

* `get host-type`:
  get host type of the server process.
  The command will put one of `storage`, `proxy`, or `archive`.

* `get vol`:
  get volume list.

* `get pid`:
  get server process process id.


============================================

For `walb-storage`,

* `get is-overflow` <VOLUME>:
  check a volume is overflow or not. It will put 0 or 1 string.
  1 means overflow.

* `get uuid`:
  get storage (wdev) uuid.

* `get ts-delta`:
  get difference of timestamps of the latest snapshot between the self host and the primary backup host for the all volumes.
  This is useful to monitor continuous backup is running correctly.
  The almost output data are the same as `walb-archive` except `kind`. `kind` must be 'storage'.


============================================

For `walb-proxy`,

* `get is-wdiff-send-error` <VOLUME> <ARCHIVE_ID>:
  check wdiff-transfer is stopped due to error. It will put 0 or 1 string.
  1 means stopped.

============================================

For `walb-archive`,

* `get diff` <VOLUME> [<GID0> [<GID1>]]:
  get diff list.
  <GID0> and <GID1> means target gid range.

* `get applicable-diff` <VOLUME> [<GID>]:
  get applicable diff list (slower than `get diff` command).
  <GID> means upper limit.

* `get total-diff-size` <VOLUME> [<GID0> [<GID1>]]:
  get total wdiff size [byte].
  This is sum of size of target wdiff files.
  You can specify gid range.

* `get num-diff` <VOLUME> [<GID0> [<GID1>]]:
  get number of wdiff files.
  You can specify gid range.

* `get exists-diff` <VOLUME> <GID0B> <GID0E> <GID1B> <GID1E>:
  <GID0B>, <GID0E> is dirty snapshot specifier of wdiff startpoint.
  <GID1B>, <GID1E> is dirty snapshot specifier of wdiff endpoint.

* `get exists-base-image` <VOLUME>:
  check existance of the base image a volume. It will put 0 or 1.
  1 means existance.

* `get num-action` <VOLUME> <ACTION>:
  get number of running actions.
  <ACTION> is one of `Merge`, `Apply`, `Restore`, `ReplSyncAsClient`, `Resize`.

* `get restored` <VOLUME>:
  get restored snapshots.

* `get restorable` <VOLUME> [all]:
  get restorable snapshots.
  Specify `all` to get implicit snapshots also.

* `get uuid` <VOLUME>:
  get storage uuid got from `walb-storage` at full-bkp/hash-bkp.

* `get archive-uuid` <VOLUME>:
  get archive uuid. This is identifier of archive data.
  Two `walb-archive` servers must have the same archive uuid to execute
  hash/diff replication.

* `get base` <VOLUME>:
  get base state (MetaState) of a volume.

* `get base-all`:
  get base state (MetaState) of the all volumes.
  Output format is LTSV. `name` is volume name, `metastate` is MetaState string of the base image.

* `get vol-size` <VOLUME>:
  get volume size [logical block].

* `get progress` <VOLUME>:
  get progress size of the running task for a volume [logical block].
  The task is one of full backup server, hash backup server,
  full replication server, resync replication server, hash replication server.
  There is at most one task concurrently.

* `get ts-delta`:
  get difference of timestamps of the latest snapshot between the self host and another hosts for the all volumes.
  This is useful to monitor replication is running correctly.
  Output format is LTSV.
  `name` is volume name. `dest` is repication destination.
  `kind` and `dest_kind` must be 'archive'.
  `gid` and `ts` are the latest snapshot information of the host: generation id and timestamp.
  `dest_gid` and `ts` are the latest snapshot information of the destination host.
  `ts_delta` is the difference of `ts` and `dest_ts` in seconds.


## SEE ALSO

wdevc(1), walb-storage(1), walb-proxy(1), walb-archive(1)


binsrc/walbc.cpp
/**
 * @file
 * @brief WalB controller tool.
 */
#include "cybozu/socket.hpp"
#include "cybozu/option.hpp"
#include "util.hpp"
#include "net_util.hpp"
#include "controller.hpp"
#include "walb_util.hpp"
#include "command_param_parser.hpp"
#include "version.hpp"

using namespace walb;

void setupOpt(cybozu::Option& opt, const char *p = "opt...", const char *help = "")
{
    static StrVec sv;
    opt.appendParamVec(&sv, p, help);
}
void setupStrVec(cybozu::Option& opt)
{
    static StrVec sv;
    opt.appendParamVec(&sv, "opt...");
}
void setupVolId(cybozu::Option& opt)
{
    static std::string volId;
    opt.appendParam(&volId, "volId");
}
void setupInitVol(cybozu::Option& opt)
{
    setupVolId(opt);
    static std::string wdevPath;
    opt.appendParamOpt(&wdevPath, "", "wdevPath");
}
void setupResetVol(cybozu::Option& opt)
{
    setupVolId(opt);
    static uint64_t gid;
    opt.appendParamOpt(&gid, 0, "gid");
}
void setupBkp(cybozu::Option& opt)
{
    setupVolId(opt);
    static uint64_t size;
    opt.appendParamOpt(&size, 0, "bulkSize");
}
void setupVolIdGid(cybozu::Option& opt)
{
    setupVolId(opt);
    static uint64_t gid;
    opt.appendParam(&gid, "gid");
}
void setupVolIdLsid(cybozu::Option& opt)
{
    setupVolId(opt);
    static uint64_t lsid;
    opt.appendParam(&lsid, "lsid");
}
void setupStart(cybozu::Option& opt)
{
    setupVolId(opt);
    static std::string target;
    opt.appendParamOpt(&target, "", "target");
}
void setupStop(cybozu::Option& opt)
{
    setupVolId(opt);
    static std::string type;
    opt.appendParamOpt(&type, "", "(opt)", ": graceful|force|empty");
}
void setupArchiveInfo(cybozu::Option& opt)
{
    setupVolId(opt);
    static StrVec sv;
    const char *help =
        "get/delete <voldId> <archiveId>\n"
        "  add/update <volId> <archiveId> <addr>:<port> <cmprType>:<cmprLevel>:<cmprNumCPU> <wdiffSendDelaySec>";
    opt.appendParamVec(&sv, "<subcmd>", help);
}
void setupReplicate(cybozu::Option& opt)
{
    setupVolId(opt);
    setupOpt(opt, "(opt)", ": ('gid' gid | 'size' sizeMb) (addr:port) (doResync) (dontMerge) (compression) (maxDiffSize) (bulkSize)");
}
void setupMerge(cybozu::Option& opt)
{
    setupVolIdGid(opt);
    setupOpt(opt, "opt", "(gidB) ('size' maxSize) | ('gid' gidE)");
}
void setupResize(cybozu::Option& opt)
{
    setupVolId(opt);
    static uint64_t size;
    static std::string zero;
    opt.appendParam(&size, "size");
    opt.appendParamOpt(&zero, "", "(opt)", ": zeroclear");
}
void setupKick(cybozu::Option& opt)
{
    setupOpt(opt, "(volId) (archiveName)");
}
void setupSetFullScanBps(cybozu::Option& opt)
{
    static uint64_t size;
    opt.appendParam(&size, "maxFullScanBps", "max full-scan throughput [bytes/sec] (0 means unlimited)");
}
void setupVirtualFullScan(cybozu::Option& opt)
{
    setupVolIdGid(opt);
    setupOpt(opt, "(bulk size) (scanning size)");
}
void setupVirtualFullScanCmd(cybozu::Option& opt)
{
    static std::string devPath;
    opt.appendParam(&devPath, "devPath", ": specify 'stdout' to put image to stdout.");
    setupVirtualFullScan(opt);
}
void setupUuid(cybozu::Option& opt)
{
    setupVolId(opt);
    static std::string uuid;
    opt.appendParam(&uuid, "uuid");
}
void setupSetState(cybozu::Option& opt)
{
    setupVolId(opt);
    static std::string state;
    opt.appendParam(&state, "state");
}
void setupShutdown(cybozu::Option& opt)
{
    static std::string force;
    opt.appendParamOpt(&force, "", "(opt)", ": force");
}

void setupDisableSnapshot(cybozu::Option& opt)
{
    setupVolId(opt);
    setupStrVec(opt);
}

void setupEnableSnapshot(cybozu::Option& opt)
{
    setupVolId(opt);
    setupStrVec(opt);
}

void setupGet(cybozu::Option& opt)
{
    std::string usage;
    usage += "usage: get NAME ARGUMENTS...\n";
    usage += "NAME list:\n";
    size_t max = 0;
    for (const protocol::GetCommandInfoMap::value_type &p : getGetCommandInfoMap()) {
        max = std::max(max, p.first.size());
    }
    max++;
    for (const protocol::GetCommandInfoMap::value_type &p : getGetCommandInfoMap()) {
        usage += cybozu::util::formatString("  %-*s%s\n", max, p.first.c_str(), p.second.helpMsg.c_str());
    }
    opt.setUsage(usage);
    static StrVec sv;
    opt.appendParamVec(&sv, "");
}

struct CommandInfo {
    std::string name;
    protocol::ClientHandler handler;
    void (*setup)(cybozu::Option&);
    void (*verify)(const StrVec&);
    const char *help;
};
const CommandInfo g_cmdTbl[] = {
    { statusCN, c2xGetStrVecClient, setupStrVec, verifyVolIdOrAllParam, "print human-readable status." },
    { initVolCN, c2xInitVolClient, setupInitVol, verifyInitVolParam, "initialize a volume." },
    { clearVolCN, c2xClearVolClient, setupVolId, verifyVolIdParam, "clear a volume." },
    { resetVolCN, c2xResetVolClient, setupResetVol, verifyResetVolParam, "reset a volume in a server." },
    { fullBkpCN, c2sBackupClient, setupBkp, verifyBackupParam, "execute full-backup." },
    { hashBkpCN, c2sBackupClient, setupBkp, verifyBackupParam, "execute hash-backup." },
    { restoreCN, c2aRestoreClient, setupVolIdGid, verifyRestoreParam, "restore a volume in an archive server." },
    { delRestoredCN, c2aDelRestoredClient, setupVolIdGid, verifyDelRestoredParam, "delete a restored volume." },
    { delColdCN, c2aDelColdClient, setupVolIdGid, verifyDelColdParam, "delete a cold volume." },
    { startCN, c2xStartClient, setupStart, verifyStartParam, "start a volume in a server." },
    { stopCN, c2xStopClient, setupStop, verifyStopParam, "stop a volume in a server." },
    { archiveInfoCN, c2pArchiveInfoClient, setupArchiveInfo, verifyArchiveInfoParam, "control archive information in a proxy." },
    { snapshotCN, c2sSnapshotClient, setupVolId, verifyVolIdParam, "take a snapshot in a storage." },
    { disableSnapshotCN, c2aDisableSnapshot, setupDisableSnapshot, verifyChangeSnapshotParam, "disable a snapshot in an archive." },
    { enableSnapshotCN, c2aEnableSnapshot, setupEnableSnapshot, verifyChangeSnapshotParam, "enable a snapshot in an archive." },
    { replicateCN, c2aReplicateClient, setupReplicate, verifyReplicateParam, "replicate a volume from an archive to another archive." },
    { applyCN, c2aApplyClient, setupVolIdGid, verifyApplyParam, "apply old wdiff files to the base image of a volume in an archive." },
    { mergeCN, c2aMergeClient, setupMerge, verifyMergeParam, "merge wdiff files for a volume in an archive." },
    { resizeCN, c2xResizeClient, setupResize, verifyResizeParam, "resize a volume in a storage or an archive." },
    { kickCN, c2xKickClient, setupKick, verifyKickParam, "kick background tasks if necessary." },
    { setFullScanBpsCN, c2sSetFullScanBpsClient, setupSetFullScanBps, verifySetFullScanBps, "set max full scan bytes per second parameter." },
    { blockHashCN, c2aBlockHashClient, setupVirtualFullScan, verifyVirtualFullScanParam, "calculate block hash of a volume in an archive." },
    { virtualFullScanCN, c2aVirtualFullScanClient, setupVirtualFullScanCmd, verifyVirtualFullScanCmdParam, "virtual full scan of a volume in an archive." },
    { getCN, c2xGetClient, setupGet, verifyNoneParam, "get some information from a server." },
    { execCN, c2xGetStrVecClient, setupStrVec, verifyNoneParam, "execute a command-line at a server's side." },
    { shutdownCN, protocol::shutdownClient, setupShutdown, verifyShutdownParam, "shutdown a server process." },
    { gcDiffCN, c2aGarbageCollectDiffClient, setupVolId, verifyVolIdParam, "garbage collect diffs." },
    { dbgReloadMetadataCN, c2aReloadMetadataClient, setupVolId, verifyVolIdParam, "reload metadata of a volue in an archive (for debug)." },
    { dbgSetUuidCN, c2aSetUuidClient, setupUuid, verifySetUuidParam, "set uuid for a volume in an archive (for debug)." },
    { dbgSetStateCN, c2aSetStateClient, setupSetState, verifySetStateParam, "set state for a volume in an archive (for debug)." },
    { dbgSetBaseCN, c2aSetBaseClient, setupSetState, verifySetBaseParam, "set base(meta-state) for a volume in an archive (for debug)." },
    { dbgDumpLogpackHeaderCN, c2sDumpLogpackHeaderClient, setupVolIdLsid, verifyDumpLogpackHeader, "dump a logpack header block(for debug)." },
#ifndef NDEBUG
    { debugCN, c2xDebugClient, setupStrVec, verifyNoneParam, "debug command (this is empty command for release)" },
#endif
};

const CommandInfo* getCommand(const std::string& cmd)
{
    for (const CommandInfo& ci : g_cmdTbl) {
        if (cmd == ci.name) return &ci;
    }
    return nullptr;
}

struct Option
{
    cybozu::Option opt1;
    cybozu::Option opt2;
    const CommandInfo *pci;
    std::string addr;
    uint16_t port;
    std::string cmd;
    StrVec params;
    std::string ctrlId;
    bool isDebug;
    size_t socketTimeout;
    void setup1stOption() {
        const std::string desc = util::getDescription("walb server controller");
        opt1.setDescription(desc);
        size_t maxLen = 0;
        for (const CommandInfo& ci : g_cmdTbl) {
            maxLen = std::max(maxLen, ci.name.size());
        }
        std::string usage =
            "usage: walbc [<opt1>] <command> [<args>]\n\n"
            "Command list:\n";
        maxLen++;
        bool first = true;
        for (const CommandInfo& ci : g_cmdTbl) {
            if (first) {
                opt1.setDelimiter(ci.name, &params);
                first = false;
            } else {
                opt1.appendDelimiter(ci.name);
            }
            usage += cybozu::util::formatString("  %-*s%s\n", maxLen, ci.name.c_str(), ci.help);
        }
        opt1.setUsage(usage, true);
        opt1.appendMust(&addr, "a", "host name or address");
        opt1.appendMust(&port, "p", "port number");
        opt1.appendParamVec(&params, "parameters", "command parameters");
        opt1.appendBoolOpt(&isDebug, "debug", "put debug message.");
        opt1.appendOpt(&socketTimeout, DEFAULT_SOCKET_TIMEOUT_SEC, "to", "Socket timeout [sec].");
        const std::string hostName = cybozu::net::getHostName();
        opt1.appendOpt(&ctrlId, hostName, "id", "controller identfier");

        opt1.appendHelp("h");
    }
    int parse1(int argc, char *argv[]) {
        setup1stOption();
        if (!opt1.parse(argc, argv)) return 0;
        const int cmdPos = opt1.getNextPositionOfDelimiter();
        if (cmdPos == 0) return 0;
        const std::string cmdName = argv[cmdPos - 1];
        pci = getCommand(cmdName);
        if (pci == nullptr) return 0;
        pci->setup(opt2);
        opt2.appendHelp("h");
        return cmdPos;
    }
    int run(int argc, char *argv[]) {
        const int cmdPos = parse1(argc, argv);
        if (cmdPos == 0) {
            opt1.usage();
            return 1;
        }
        if (pci->name == execCN) { // `exec` cmd does not parse opt2
            if (cmdPos == argc) {
                opt2.usage();
                return 1;
            }
        } else {
            if (!opt2.parse(argc, argv, cmdPos)) {
                opt2.usage();
                return 1;
            }
        }
        util::setLogSetting("-", isDebug);
        pci->verify(params);

        cybozu::Socket sock;
        util::connectWithTimeout(sock, cybozu::SocketAddr(addr, port), socketTimeout);
        std::string serverId = protocol::run1stNegotiateAsClient(sock, ctrlId, pci->name);
        ProtocolLogger logger(ctrlId, serverId);
        protocol::ClientParams clientParams(sock, logger, params);
        pci->handler(clientParams);
        return 0;
    }
};


int main(int argc, char *argv[])
try {
    Option opt;
    return opt.run(argc, argv);
} catch (std::exception &e) {
    LOGe("walbc error: %s", e.what());
    return 1;
}
